### ISSUE TITLE
[no-unnecessary-condition] False positive on class field

### ISSUE THREAD

**Issue by glen-84** [typescript-eslint/typescript-eslint#996]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": ["error", { "ignoreRhs": true }],
  }
}
```

```TS
class Thing {}

class Example {
    private test: Thing | null;

    public run(): void {
        if (this.test) { // Error here.
            console.log(this.test);
        }
    }
}
```

**Expected Result**

No error.

**Actual Result**

Error: Unnecessary conditional, value is always truthy.

**Additional Info**

n/a

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.3.0` |
| `@typescript-eslint/parser`        | `2.3.0` |
| `TypeScript`                       | `3.5.3` |
| `ESLint`                           | `6.1.0` |
| `node`                             | `10.16.3` |
| `npm`                              | `6.9.0` |

**Comment by bradzacher** [comment]
I'm unable to reproduce this against master.
Do you have strictNullChecks on?

If you don't - then the actual type of `this.test` degrades to `Thing` - hence the rule flags the condition as unnecessary.

**Comment by glen-84** [comment]
I'll try to test this again on Friday.

**Comment by glen-84** [comment]
`strictNullChecks` was enabled, but I think that I had forgotten to reload VS Code, so the ESLint extension may have been running with `strictNullChecks` disabled.

Having said that, with `strictNullChecks` _disabled_, shouldn't:

```ts
class Thing {}

class Example {
    private test: Thing;

    public run(): void {
        if (this.test) {
            // Error here.
            console.log(this.test);
        }
    }
}
```

... be fine, if `null` and `undefined` are [in the domain of every type](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html)?

It's possible that I am misunderstanding something, but you'd think that code like the above would be common in a world without `strictNullChecks`.

**Comment by bradzacher** [comment]
You're correct in your understanding of what happens when it's turned off.

But incorrect in the understanding of how the typechecker reports types.
When strict mode is off, the types reported by the checker API aren't any different to when it's on.

So when we query if the type is nullable etc, the checker api just tells us that it's of type `Thing` - it's up to us to implement the non-strict logic.

We've done it for a few rules, so we need to add it where appropriate.

**Comment by glen-84** [comment]
Ya, I knew that it was being reported as `Thing` (by hovering over the symbol in VS Code), I just wasn't sure whether the APIs provided additional metadata, or whether it had to be handled manually.

**Comment by Stephen2** [comment]
Am I seeing the same thing here?
![image](https://user-images.githubusercontent.com/4276422/66012752-356ea780-e47d-11e9-86f6-7c5861959174.png)

![image](https://user-images.githubusercontent.com/4276422/66012756-3b648880-e47d-11e9-938a-98ea9f185b86.png)

**Comment by Stephen2** [comment]
Seems inaccurate to me.  Just started happening after an upgrade, I'm gonna rollback and see what happens

**Comment by Stephen2** [comment]
Went from `2.1.0` -> `2.3.2` and this popped up a bunch.  Downgraded step by step.  `2.2.0+` saw these false positives appear, so I had to go back to `2.1.0`

**Comment by Stephen2** [comment]
Confirmed, going from `2.1.0` -> `2.2.0` causes these false positives for me.

I even rolled back `typescript` version to `3.5.3` because I was getting the giant warning message about being outside estree supported version

But still no luck.

**Comment by Stephen2** [comment]
blerg... I tried hacking up `node-modules` to figure out what might be the cause, but I didn't understand it :(

Sorry... I tried to help

**Comment by glen-84** [comment]
@Stephen2 Are you using `"ignoreRhs": true`?

**Comment by Stephen2** [comment]
I'm not

**Comment by Stephen2** [comment]
Turning that on makes the list of 20 silly ones go down to 3 mostly reasonable ones 🍾 

Am I misunderstanding how the rules works?

**Comment by glen-84** [comment]
@Stephen2 #1017.

**Comment by derekparsons718** [comment]
I would love to see this rule work better when strict null checks is disabled.

Until that happens, I would suggest updating the documentation to indicate this limitation. A little note like "this rule requires strict-null-checks" could save people like me from scratching their heads so much :)

Thanks for all of your hard work!
